### PR TITLE
Adding single kernel for string to decimal casting

### PR DESCRIPTION
**Pull request by hyperbolic2346**
This adds a kernel to support casting a string to a decimal.

This is currently still WIP as there are some known edge cases from the plugin tests. I wanted to get this up to get input and decide about 22.10 vs 22.12 release.

### CODE REVIEW

**Comment by revans2** [comment]
@hyperbolic2346 you need to sign off on your changes in one of the commits.

File: src/main/cpp/tests/cast_string.cpp [hunk 1]
```diff
@@ -229,4 +231,212 @@ TYPED_TEST(StringToIntegerTests, Overflow)
...
+TEST_F(StringToDecimalTests, OverPrecise)
+{
+  auto const strings = test::strings_column_wrapper({"123456", "999999", "-123456", "-999999"});
+  strings_column_view scv{strings};
+
+  auto const result =
+    spark_rapids_jni::string_to_decimal(5, 0, scv, false, rmm::cuda_stream_default);
+
+  test::fixed_point_column_wrapper<int32_t> expected(
+    {12346, 100000, -12346, -100000}, {0, 0, 0, 0}, numeric::scale_type{0});
```
Comment (revans2) [inline src/main/cpp/tests/cast_string.cpp:260]: nit: This test is a little confusing. With a precision of 5 and a scale of 0 all of the numbers overflow and you get null back.  All of that is great. The confusing part is that you don't have an XXX like in other code that indicates it is a NULL placeholder so I don't spend a few mins trying to understand why these values look like they are rounded almost correctly.

File: src/main/cpp/tests/cast_string.cpp [hunk 2]
```diff
@@ -229,4 +231,212 @@ TYPED_TEST(StringToIntegerTests, Overflow)
...
+
+  test::print(result->view());
```
Comment (revans2) [inline src/main/cpp/tests/cast_string.cpp:262]: nit: do we still need/want this print?

File: src/main/cpp/src/cast_string.cu [hunk 3]
```diff
@@ -271,4 +756,37 @@ std::unique_ptr<column> string_to_integer(data_type dtype,
...
+std::unique_ptr<column> string_to_decimal(int32_t precision,
+                                          int32_t scale,
+                                          strings_column_view const& string_col,
+                                          bool ansi_mode,
+                                          rmm::cuda_stream_view stream,
+                                          rmm::mr::device_memory_resource* mr)
+{
+  data_type dtype = [precision, scale]() {
+    if (precision <= cuda::std::numeric_limits<int32_t>::digits10)
+      return data_type(type_id::DECIMAL32, -scale);
```
Comment (revans2) [inline src/main/cpp/src/cast_string.cu:780]: nit: can we have the scale passed in be the CUDF scale and not the java/scala/spark scale?  This is for consistency. All of the other CUDF java APIs take the CUDF scale.

File: src/main/cpp/tests/cast_string.cpp [hunk 4]
```diff
@@ -229,4 +231,212 @@ TYPED_TEST(StringToIntegerTests, Overflow)
...
+TEST_F(StringToDecimalTests, Rounding)
+{
+  auto const strings = test::strings_column_wrapper({"1.23456", "9.99999", "-1.23456", "-9.99999"});
+  strings_column_view scv{strings};
+
+  auto const result =
+    spark_rapids_jni::string_to_decimal(5, 4, scv, false, rmm::cuda_stream_default);
+
+  test::fixed_point_column_wrapper<int32_t> expected(
+    {12346, 100000, -12346, -100000}, {1, 1, 1, 1}, numeric::scale_type{-4});
```
Comment (revans2) [inline src/main/cpp/tests/cast_string.cpp:276]: So it is expected that the java code still needs to do an overflow check on the output of this? The second and fourth numbers technically do not fit in a precision of 5. The precision is 6 after rounding. If this is expected that is fine, we just need to make sure that we have documented it. Just because it is not what I would have expected.

File: src/main/cpp/tests/cast_string.cpp [hunk 5]
```diff
@@ -229,4 +231,212 @@ TYPED_TEST(StringToIntegerTests, Overflow)
...
+    strings_column_view scv{str};
+
+    auto const result =
+      spark_rapids_jni::string_to_decimal(6, 6, scv, false, rmm::cuda_stream_default);
+    test::fixed_point_column_wrapper<int32_t> expected({0}, {0}, numeric::scale_type{-6});
+
+    test::print(result->view());
+
+    CUDF_TEST_EXPECT_COLUMNS_EQUAL(result->view(), expected);
+  }
```
Comment (revans2) [inline src/main/cpp/tests/cast_string.cpp:441]: Does this support values with a negative java/scala/spark scale?  If so we are missing tests for this. If not then we should document it.  Support for it is turned off by default in Spark, so I don't think it is a big deal for us to not support it right now.

File: src/main/cpp/tests/cast_string.cpp (hunk 1 above)
Comment (hyperbolic2346) [reply src/main/cpp/tests/cast_string.cpp:260]: I agree and this is a result of copy/paste. I have changed this to 0's, but would you prefer a null variable name instead?

File: src/main/cpp/src/cast_string.cu (hunk 3 above)
Comment (hyperbolic2346) [reply src/main/cpp/src/cast_string.cu:780]: This is actually helpful for understanding the code as well. I'll make this change.

File: src/main/cpp/tests/cast_string.cpp (hunk 1 above)
Comment (revans2) [reply src/main/cpp/tests/cast_string.cpp:260]: That is good enough for me.

File: src/main/cpp/tests/cast_string.cpp (hunk 4 above)
Comment (hyperbolic2346) [reply src/main/cpp/tests/cast_string.cpp:276]: Not at all. My goal is to match `spark-shell` in the handling. I thought I was doing that, but I see that spark will return null for these tests. I will update the code to ensure it matches the output of spark.

File: src/main/cpp/tests/cast_string.cpp (hunk 5 above)
Comment (hyperbolic2346) [reply src/main/cpp/tests/cast_string.cpp:441]: I believe it will, but I haven't tested it. I need to do that and then decide about fixing or adding comments about it. If this doesn't support negative scale, I would need to adjust the plugin to still be able to call the original code in those cases, right?

Comment (revans2) [reply src/main/cpp/tests/cast_string.cpp:441]: Yes

File: src/main/cpp/tests/cast_string.cpp (hunk 4 above)
Comment (hyperbolic2346) [reply src/main/cpp/tests/cast_string.cpp]: Updated these to match spark-shell and updated the code to produce valid results.

File: src/main/cpp/tests/cast_string.cpp (hunk 5 above)
Comment (hyperbolic2346) [reply src/main/cpp/tests/cast_string.cpp:426]: I added a test for negative scales, but I'm not 100% these match spark. I tried to enable negative scales, but it still complained. Need to figure this out so I can verify negative scales are working as intended.
```
scala> spark.conf.set("spark.sql.legacy.allowNegativeScaleOfDecimal", true)

scala> spark.sql("""select cast("1.234e-1" as decimal(6,-2))""").show
org.apache.spark.sql.catalyst.parser.ParseException:
extraneous input '-' expecting INTEGER_VALUE(line 1, pos 36)

== SQL ==
select cast("1.234e-1" as decimal(6,-2))
------------------------------------^^^

  at org.apache.spark.sql.catalyst.parser.ParseException.withCommand(ParseDriver.scala:266)
  at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parse(ParseDriver.scala:127)
  at org.apache.spark.sql.execution.SparkSqlParser.parse(SparkSqlParser.scala:51)
  at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parsePlan(ParseDriver.scala:77)
  at org.apache.spark.sql.SparkSession.$anonfun$sql$2(SparkSession.scala:616)
  at org.apache.spark.sql.catalyst.QueryPlanningTracker.measurePhase(QueryPlanningTracker.scala:111)
  at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:616)
  at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:775)
  at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:613)
  ... 47 elided
```

Comment (hyperbolic2346) [reply src/main/cpp/tests/cast_string.cpp:459]: Finished testing positive scale values thanks to Bobby's help:
```
import org.apache.spark.sql.types._
spark.conf.set("spark.sql.legacy.allowNegativeScaleOfDecimal", "true")
Seq("-7.691546687437984E16").toDF().withColumn("dec", col("value").cast(DecimalType(10, -1))).show()
```

**Comment by hyperbolic2346** [comment]
build

**Comment by hyperbolic2346** [comment]
build

**Comment by hyperbolic2346** [comment]
build

File: src/main/cpp/src/cast_string.cu [hunk 6]
```diff
@@ -37,6 +37,12 @@ namespace detail {
 
 constexpr auto NUM_THREADS{256};
 
+/**
+ * @brief identifies if a character is whitespace
```
Comment (ttnghia) [inline src/main/cpp/src/cast_string.cu:41]:
```suggestion
 * @brief Identify if a character is whitespace.
```

File: src/main/cpp/src/cast_string.cu [hunk 7]
```diff
@@ -48,8 +54,126 @@ constexpr bool is_whitespace(char const chr)
...
+
+template <typename T, std::enable_if_t<not cuda::std::is_signed_v<T>>* = nullptr>
+constexpr T __device__ generic_abs(T value)
+{
+  return value;
+}
+
+/**
+ * @brief determines if overflow will occur when multiplying
+ * a value by 10
```
Comment (ttnghia) [inline src/main/cpp/src/cast_string.cu:71]:
```suggestion
 * @brief Determine if overflow will occur when multiplying a value by 10.
```

File: src/main/cpp/src/cast_string.cu [hunk 8]
```diff
@@ -48,8 +54,126 @@ constexpr bool is_whitespace(char const chr)
...
+bool __device__ will_overflow(T const val, bool adding)
+{
+  if constexpr (std::is_signed_v<T>) {
+    if (!adding) {
+      auto constexpr minval = cuda::std::numeric_limits<T>::min() / 10;
+      if (val < minval) {
+        // overflow
+        return true;
+      }
+      return false;
```
Comment (ttnghia) [inline src/main/cpp/src/cast_string.cu:88]:
```suggestion
      return val < minval;
```

File: src/main/cpp/src/cast_string.cu (hunk 7 above)
Comment (ttnghia) [reply src/main/cpp/src/cast_string.cu:71]: This also needs to clarify why do we need the `adding` parameter.

File: src/main/cpp/src/cast_string.cu [hunk 9]
```diff
@@ -48,8 +54,126 @@ constexpr bool is_whitespace(char const chr)
...
+    }
+  }
+
+  auto constexpr maxval = cuda::std::numeric_limits<T>::max() / 10;
+  if (val > maxval) {
+    // overflow
+    return true;
+  }
+
+  return false;
```
Comment (ttnghia) [inline src/main/cpp/src/cast_string.cu:98]:
```suggestion
  return val > maxval;
```

File: src/main/cpp/src/cast_string.cu [hunk 10]
```diff
@@ -48,8 +54,126 @@ constexpr bool is_whitespace(char const chr)
...
+}
+
+/**
+ * @brief determines if overflow will occur when adding or subtracting values
```
Comment (ttnghia) [inline src/main/cpp/src/cast_string.cu:102]:
```suggestion
 * @brief Determine if overflow will occur when adding or subtracting values.
```

File: src/main/cpp/src/cast_string.cu [hunk 11]
```diff
@@ -48,8 +54,126 @@ constexpr bool is_whitespace(char const chr)
...
+bool __device__ will_overflow(T const lhs, T const rhs, bool adding)
+{
+  if constexpr (std::is_signed_v<T>) {
+    if (!adding) {
+      auto const min_val = cuda::std::numeric_limits<T>::min() + rhs;
+      if (lhs < min_val) {
+        // overflow
+        return true;
+      }
+      return false;
```
Comment (ttnghia) [inline src/main/cpp/src/cast_string.cu:120]: Simplify code similar to above.

File: src/main/cpp/src/cast_string.cu [hunk 12]
```diff
@@ -48,8 +54,126 @@ constexpr bool is_whitespace(char const chr)
...
+ *
+ * @tparam T type of the input data
+ * @param first_value true if this is the first value added
+ * @param current_val current computed value
+ * @param new_digit digit to append to the computed value
+ * @param adding true if adding, false if subtracting
+ * @return true if success, false if overflow
+ */
+template <typename T>
+bool __device__ process_value(bool first_value, T& current_val, T const new_digit, bool adding)
```
Comment (ttnghia) [inline src/main/cpp/src/cast_string.cu:144]: We should not use ref in input parameter. Instead, return a pair of `<ret, bool>`.

Comment (hyperbolic2346) [reply src/main/cpp/src/cast_string.cu]: Updated to fix this, but I went with `<bool, ret>` for no particular reason.

**Comment by hyperbolic2346** [comment]
build

File: src/main/cpp/src/cast_string.cu [hunk 13]
```diff
@@ -115,48 +221,350 @@ __global__ void string_to_integer_kernel(T* out,
...
+template <typename T>
+__global__ void string_to_decimal_kernel(T* out,
+                                         bitmask_type* validity,
+                                         const char* const chars,
+                                         offset_type const* offsets,
+                                         bitmask_type const* incoming_null_mask,
+                                         size_type num_rows,
+                                         int32_t scale,
+                                         int32_t precision,
+                                         bool ansi_mode)
```
Comment (ttnghia) [inline src/main/cpp/src/cast_string.cu:393]: I don't see `ansi_mode` is used anywhere in this function?

File: src/main/cpp/src/cast_string.cu [hunk 14]
```diff
@@ -176,7 +584,59 @@ struct row_valid_fn {
...
+                                             thrust::make_counting_iterator(col.size()),
+                                             row_valid_fn{col.null_mask(), source_col.null_mask()});
+
+    offset_type string_bounds[2];
+    cudaMemcpyAsync(&string_bounds,
+                    &source_col.offsets().data<offset_type>()[*first_error],
+                    sizeof(offset_type) * 2,
+                    cudaMemcpyDeviceToHost,
+                    stream.value());
+    stream.synchronize();
```
Comment (ttnghia) [inline src/main/cpp/src/cast_string.cu:613]: If we need copy then sync stream then why not just use `cudaMemcpy` (sync copy)?

File: src/main/cpp/src/cast_string.cu [hunk 15]
```diff
@@ -176,7 +584,59 @@ struct row_valid_fn {
...
+ */
+void validate_ansi_column(column_view const& col,
+                          strings_column_view const& source_col,
+                          rmm::cuda_stream_view stream)
+{
+  auto const num_nulls      = col.null_count();
+  auto const incoming_nulls = source_col.null_count();
+  auto const num_errors     = num_nulls - incoming_nulls;
+  if (num_errors > 0) {
+    auto const first_error = thrust::find_if(rmm::exec_policy(stream),
```
Comment (ttnghia) [inline src/main/cpp/src/cast_string.cu:602]: `thrust::find_if` is known to have performance issue. Use `count_if(...) > 0`.

Comment (ttnghia) [inline src/main/cpp/src/cast_string.cu:602]: What if `first_error` iter is out of bound?

File: src/main/cpp/src/cast_string.cu [hunk 16]
```diff
@@ -202,42 +662,14 @@ struct string_to_integer_impl {
...
-                        &string_col.chars().data<char const>()[string_bounds[0]],
-                        string_bounds[1] - string_bounds[0],
-                        cudaMemcpyDeviceToHost,
-                        stream.value());
-        stream.synchronize();
-
-        throw cast_error(*first_error, dest);
-      }
-    }
+    if (ansi_mode) { validate_ansi_column(col->view(), string_col, stream); }
```
Comment (ttnghia) [inline src/main/cpp/src/cast_string.cu:665]: Can we validate while casting, instead of calling to a separate kernel? We can make `ansi_mode` a template parameter to reduce performance impact.

File: src/main/cpp/src/cast_string.cu [hunk 17]
```diff
@@ -248,6 +680,70 @@ struct string_to_integer_impl {
...
+    using Type = device_storage_type_t<T>;
+
+    rmm::device_uvector<Type> data(string_col.size(), stream, mr);
+    auto const num_words = bitmask_allocation_size_bytes(string_col.size()) / sizeof(bitmask_type);
+    rmm::device_uvector<bitmask_type> null_mask(num_words, stream, mr);
+
+    dim3 const blocks(util::div_rounding_up_unsafe(string_col.size(), detail::NUM_THREADS));
+    dim3 const threads{detail::NUM_THREADS};
+
+    detail::string_to_decimal_kernel<<<blocks, threads, 0, stream.value()>>>(
```
Comment (ttnghia) [inline src/main/cpp/src/cast_string.cu:713]: Can we maximize code share between `string_to_integer_kernel` and `string_to_decimal_kernel`? Like extracting some helper functions that can be called from both places.

File: src/main/cpp/src/cast_string.cu (hunk 15 above)
Comment (hyperbolic2346) [reply src/main/cpp/src/cast_string.cu:602]: if `num_errors` > 0 then `first_error` will be valid. I can check this anyway, but it would be a sanity check.

File: src/main/cpp/src/cast_string.cu (hunk 16 above)
Comment (hyperbolic2346) [reply src/main/cpp/src/cast_string.cu:665]: I was torn on this myself, but I assume the majority of cases will be non-ansi and as a result, I didn't want to impact performance of the common path. If we have an ansi mode request and there are errors, I'm not concerned about the kernel launch for it. It does introduce possibly 2 kernel launches to determine if we have errors though when we count the nulls, but I chalked that up as the cost of admission.

If we pushed this content into the kernel itself, we would lose some of the commonality for it. Each thread would need to either atomic set a bit somewhere or atomic increment or we each row would have a mask similar to the null mask. The problem there is the extra memory required over the common case. This was the least-intrusive way to address this I could find.

File: src/main/cpp/src/cast_string.cu (hunk 17 above)
Comment (hyperbolic2346) [reply src/main/cpp/src/cast_string.cu:713]: That is 100% the goal here and I did pull out some common functions like `will_overflow`, `is_whitespace`, and `process_value`. I believe there is more sharing going on here and we can pull more out. My plan was to continue to factor out common code as more casts are added. There is a list we are going down for these.

**Comment by ttnghia** [comment]
LGTM now (may have more nitpick later) but I would like to stress test this with more integration tests (i.e., various random input) in https://github.com/NVIDIA/spark-rapids/pull/6616. If we can run a lot of tests before merging this then we will not have to go back and fix bugs showing up later.

**Comment by hyperbolic2346** [comment]
This has been tested in `spark-rapids` and passes all tests now. The tests in this project are issues that were found in testing and known edges. The real extensive testing is done in the plugin. Are you asking for more tests to be added there?

**Comment by revans2** [comment]
build
